Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx

```diff
@@ -0,0 +1,304 @@
+#endif
+
+  void process(aod::HfCandCascade const& candidates,
+               aod::BigTracksMC const& tracks,
+               aod::McParticles const& particlesMC)
+  {
+    int8_t sign = 0;
+    std::vector<int> arrDaughLcIndex;
+    std::array<int, 3> arrDaughLcPDG;
+    std::array<int, 3> arrDaughLcPDGRef = {2212, 211, -211};
+
+    // Match reconstructed candidates.
+    for (auto& candidate : candidates) {
+      auto arrayDaughtersV0 = array{candidate.posTrack_as<aod::BigTracksMC>(), candidate.negTrack_as<aod::BigTracksMC>()};
+      auto arrayDaughtersLc = array{candidate.index0_as<aod::BigTracksMC>(), candidate.posTrack_as<aod::BigTracksMC>(), candidate.negTrack_as<aod::BigTracksMC>()};
+
+      // First we check the K0s
+      LOG(DEBUG) << "\n";
+      LOG(DEBUG) << "Checking MC for candidate!";
+      LOG(DEBUG) << "Looking for K0s";
+      auto labelPos = candidate.posTrack_as<aod::BigTracksMC>().mcParticleId();
+      auto labelNeg = candidate.negTrack_as<aod::BigTracksMC>().mcParticleId();
+      auto protonLabel = candidate.index0_as<aod::BigTracksMC>().mcParticleId();
+
+#ifdef MY_DEBUG
+      bool isLc = isLcK0SpFunc(protonLabel, labelPos, labelNeg, v_labelProton, v_labelK0Spos, v_labelK0Sneg);
+      bool isK0SfromLc = isK0SfromLcFunc(labelPos, labelNeg, v_labelK0Spos, v_labelK0Sneg);
+#endif
+      MY_DEBUG_MSG(isK0SfromLc, LOG(INFO) << "correct K0S in the Lc daughters: posTrack --> " << labelPos << ", negTrack --> " << labelNeg);
+
+      //if (isLc) {
+      RecoDecay::getMatchedMCRec(particlesMC, arrayDaughtersV0, kK0Short, array{+kPiPlus, -kPiPlus}, true, &sign, 1); // does it matter the "acceptAntiParticle" in the K0s case? In principle, there is no anti-K0s
+
+      if (sign != 0) { // we have already positively checked the K0s
+        // then we check the Lc
+        MY_DEBUG_MSG(sign, LOG(INFO) << "K0S was correct! now we check the Lc");
+        auto labelProton = candidate.index0_as<aod::BigTracksMC>().mcParticleId();
+        MY_DEBUG_MSG(sign, LOG(INFO) << "label proton = " << labelProton);
+        RecoDecay::getMatchedMCRec(particlesMC, arrayDaughtersLc, pdg::code::kLambdaCPlus, array{+kProton, +kPiPlus, -kPiPlus}, true, &sign, 3); // 3-levels Lc --> p + K0 --> p + K0s --> p + pi+ pi-
+        MY_DEBUG_MSG(sign, LOG(INFO) << "Lc found with sign " << sign; printf("\n"));
+      }
+
+      rowMCMatchRec(sign);
+    }
+    //}
+
+    // Match generated particles.
+    for (auto& particle : particlesMC) {
+      // checking if I have a Lc --> K0S + p
+      RecoDecay::isMatchedMCGen(particlesMC, particle, pdg::code::kLambdaCPlus, array{+kProton, +kK0Short}, true, &sign, 2);
+      if (sign != 0) {
+        MY_DEBUG_MSG(sign, LOG(INFO) << "Lc in K0S p");
+        arrDaughLcIndex.clear();
+        // checking that the final daughters (decay depth = 3) are p, pi+, pi-
+        RecoDecay::getDaughters(particlesMC, particle, &arrDaughLcIndex, arrDaughLcPDGRef, 3); // best would be to check the K0S daughters
```

Comment:
You can get `arrDaughLcIndex` directly from `isMatchedMCGen`.

Reply:
I am not sure here: will it then check that it is really Lc --> K0 + p --> K0S + p --> pi+ pi- p?

Reply:
Leaving as it is for now, do you agree?